Rest client classic: revert to package private visibility for configuration methods

TomasHofman opened the pull request:
1. Use "protected" visibility for `RestClientBase.configure*` methods, to accommodate kogito extension that accesses these. Originally the visibility of these methods was package private, but was changed to private, which broke kogito extension.
2. Introduce `RestClientBuilder` factories, that allow developers to produce `RestClientBuilder` instances that are already configured according to application.properties (reusing the configuration code from classic/reactive rest-client extensions). The builders can then be further tweaked and finally used to programmatically produce rest client proxies.

Related to https://github.com/quarkusio/quarkus/issues/12486

TomasHofman committed:
Rest client classic: make configuration methods in RestClientBase protected

This is to accommodate kogito extension that accesses these methods.

gsmet commented:
Can you confirm that at some point you have a way out of this? I understand from the original issue that you need that change to be able to support multiple Quarkus versions. But if at some point your minimal version is current main, you will be able to switch to something else or there's something we missed?

Thanks.

fjtirado commented:
> > Can you confirm that at some point you have a way out of this? I understand from the original issue that you need that change to be able to support multiple Quarkus versions. But if at some point your minimal version is current main, you will be able to switch to something else or there's something we missed?
> > Thanks.
> 
> @cristianonicolai would you be able to answer @gsmet 's question?
The plan is to use configureBuilder as soon as the method is avaiable and remove the call to separate configureXXX methods
The proposal is to change the scope of these method from package to protected rather than private.

cristianonicolai commented:
@gsmet we're tied to RHBQ at this stage, once that is bumped, we can remove such code.

TomasHofman commented:
Please do not merge yet, I'm going to update the PR.

TomasHofman committed:
Add RestClientBuilder factories

These factories should make it easier for developers to obtain a
RestClientBuilder instance preconfigured according to current
application config. The builder can be further tweaked and then used to
programatically build a rest client proxy.

TomasHofman committed:
Rest client config: Remove incorrect note about the proxyAddress property being only applicable to reactive clients

fjtirado reviewed (approved):
Thanks a lot @TomasHofman

TomasHofman commented:
Cc @geoand @Sgitario if you wanted to review this.

I think in past we talked about providing developers with some way to reuse client configuration code from rest-client extensions when producing clients programmatically.

(PS. I'm aware of the formatter issues, will fix them ASAP.)

gsmet commented:
Thanks!
